### PR TITLE
python : fix SSL for older Python versions

### PR DESCRIPTION
**Pull request by xjrc**
The changes in this pull request fix SSL for older Python versions by adding a new `patch` function and by adding dependency constraints.

**NOTE**: I've only verified that these fixes work for the variant `python@2.7.12+ssl` with a build environment of `gcc%4.9.3 arch=linux-rhel7-broadwell`, so this should remain a WIP until I've more thoroughly tested the correctness of these changes. I'm planning to at least test  `python@2.7.18+ssl`, `python@3.4.X+ssl`, and `python@3.6.X+ssl` to ensure that the patch and constraints work for the advertised versions.

**EDIT**: I've now tested `python@2.7.18{+|~}ssl` and `python@3.6.8{+|~}ssl`. I skipped `python@3.4.3+ssl` because it seems like it would be redundant given that the patches in this PR work for a more recent version (i.e.  `python@3.6.8`).

### CODE REVIEW

File: var/spack/repos/builtin/packages/python/package.py [hunk 1]
```diff
@@ -202,6 +204,20 @@ def patch(self):
             r'\1setup.py\2 --no-user-cfg \3\6'
         )
 
+    @when('@:3.6.999+ssl')
+    @run_before('build')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/python/package.py]: Does @when and @run_before work in conjunction with each other? I feel like I've had problems with this in the past, you should make sure it doesn't always run or never run.

Comment (xjrc) [reply var/spack/repos/builtin/packages/python/package.py]: I retried the install with variants that should both pass and fail these `@when` conditions (specifically, `python@2.7.12+ssl` and `python@2.7.12~ssl`) and it looks like `@run_before` and `@when` don't play well together. In particular, it looks like the `@when` gets ignored and the function is always run, which explains why I didn't encounter problems before when I was running purely `+ssl` variants.

This being the case, what's the best way to run both of these patches? I don't think multiple `patch` functions will work unless they span disjoint specs (see `qt`), so is the only option to write a general `patch` function that has conditional clauses for these two different cases?

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/python/package.py]: Yeah, I would recommend a single patch function with two if statements, one for each version range.

Comment (alalazo) [reply var/spack/repos/builtin/packages/python/package.py]: See https://github.com/spack/spack/issues/12736. If we encounter a lot of cases like this we might raise the impact level of the bug.

**Comment by xjrc** [comment]
Okay, I merged the two `@run_before`/`@when` functions into a single `patch` function, combed over the per-version SSL version requirements more carefully, and did more thorough testing of the `{~|+}ssl` variant by doing `ldd` and `import ssl`/`import hashlib` tests on installed `python` binaries. I've verified that all of the following variants work when building with `gcc%4.9.3` on `arch=linux-rhel7-broadwell`:

- `@2.7.12{~|+}ssl` (the version I'm installing)
- `@2.7.18{~|+}ssl` (the latest in the `@2` series)
- `@3.6.8{~|+}ssl` (the latest in the `@3:3.6.999` series, before `--with-ssl=`)
- `@2.7.18+ssl ^openssl@1.0.2u` (compatibility testing)
- `@3.6.8+ssl ^openssl@1.0.2u` (compatibility testing)

@adamjstewart: Please let me know if there's anything else I should test or modify to get these changes merged!

**Comment by adamjstewart** [comment]
Flake8:
```
var/spack/repos/builtin/packages/python/package.py:229: [E501] line too long (81 > 79 characters)
var/spack/repos/builtin/packages/python/package.py:230: [E501] line too long (80 > 79 characters)
```

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**